airbyte-db/db-lib/src/main/java/io/airbyte/db/mongodb/MongoUtils.java

```diff
@@ -275,33 +277,39 @@ private static String transformName(final List<String> types, final String name)
 
   private static void setSubFields(final MongoCollection<Document> collection,
                                    final TreeNode<CommonField<BsonType>> parentNode,
-                                   final String pathToField) {
+                                   final String pathToField,int maxLevel) {
     final var nestedKeys = getFieldsName(collection, pathToField);
-    nestedKeys.forEach(key -> {
+    if (maxLevel<=0) return ;
+    for (String key : nestedKeys) {
+      // if empty field process them as '-'
+      if (key == null || key.length() == 0) key = "-";
       final var types = getTypes(collection, pathToField + "." + key);
       final var nestedType = getUniqueType(types);
       final var childNode = parentNode.addChild(new CommonField<>(transformName(types, key), nestedType));
       if (nestedType.equals(DOCUMENT)) {
-        setSubFields(collection, childNode, pathToField + "." + key);
+        setSubFields(collection, childNode, pathToField + "." + key, maxLevel--);
       }
-    });
+    }
   }
 
   private static List<String> getFieldsName(final MongoCollection<Document> collection) {
     return getFieldsName(collection, "$ROOT");
   }
-
   private static List<String> getFieldsName(final MongoCollection<Document> collection, final String fieldName) {
     final AggregateIterable<Document> output = collection.aggregate(Arrays.asList(
         new Document("$limit", DISCOVER_LIMIT),
         new Document("$project", new Document("arrayofkeyvalue", new Document("$objectToArray", "$" + fieldName))),
         new Document("$unwind", "$arrayofkeyvalue"),
-        new Document("$group", new Document(ID, null).append("allkeys", new Document("$addToSet", "$arrayofkeyvalue.k")))));
-    if (output.cursor().hasNext()) {
-      return (List) output.cursor().next().get("allkeys");
-    } else {
-      return Collections.emptyList();
-    }
+        new Document("$group", new Document(ID, null).append("allkeys", new Document("$addToSet", "$arrayofkeyvalue.k")))
+        ));
+    // cursor based approach for saving memory
+      List<String> allKeys = new ArrayList<>();
+      try (MongoCursor<Document> cursor = output.cursor()) {
+          while (cursor.hasNext()) {
+              allKeys.addAll((List<String>) cursor.next().get("allkeys"));
+          }
+        }
+      return allKeys;
```

Comment:
here i am returning all keys if their is nothing then it will be empty  @marcosmarxm